Services/AssessmentQuestion/examples/class.exTestPlayerGUI.php

```diff
@@ -129,68 +117,112 @@ public function submitSolution()
 	{
 		global $DIC; /* @var ILIAS\DI\Container $DIC */
 		
-		// this can also be $_REQUEST or any other future ilias post-request handler
-		$serverRequestObject; /* @var \Psr\Http\Message\ServerRequestInterface $serverRequestObject */
-		
-		$questionId = 0; // initialise with id of question that just submits
-		
 		/**
-		 * fetch possibly existing participant solution, an empty one is required otherwise
+		 * initialise with id of question to be shown
 		 */
 		
-		$participantSolution = $this->getParticipantSolution($questionId);
+		$questionUuid = $DIC->assessment()->consumer()->questionUuid('any-valid-question-uuid');
+		$revisionUuid = $DIC->assessment()->consumer()->questionUuid('any-valid-revision-uuid');
 		
 		/**
-		 * let the solution object instance harvest the submission post data
+		 * fetch possibly existing participant answer uuid,
+		 * when no participant answer exist yet,
+		 * generate a new user answer uuid
 		 */
 		
-		$participantSolution->initFromServerRequest($serverRequestObject);
+		$userAnswerUuid = $this->getParticipantAnswerUuid($questionUuid->getId());
+		
+		if( $userAnswerUuid )
+		{
+			$userAnswerUuid = $DIC->assessment()->consumer()->userAnswerUuid($userAnswerUuid);
+		}
+		else
+		{
+			$userAnswerUuid = $DIC->assessment()->consumer()->newUserAnswerUuid();
+		}
 		
 		/**
-		 * get results calculator to be used to retrieve calculated reached points
-		 * that can be stored in a test result storage managed by the test object
+		 * generate a user answer submit containing the post data
 		 */
+		$asqPlayService = $DIC->assessment()->service()->play(
+			$this->buildAsqPlayServiceSpec(), $DIC->assessment()->consumer()->questionUuid($questionUuid),$DIC->assessment()->consumer()->revisionUuid($questionUuid,$revisionUuid)
+		);
+		
+		$DIC->assessment()->service()->play()->CreateUserAnswer(
+				new UserAnswerSubmitContract(
+					$DIC->assessment()->consumer()->UserAnswerUuid(
+						new PostDataFromServerRequest($request)->get('user_answer_uuid')
+					),
+                    $DIC->assessment()->consumer()->questionUuid(
+                        new PostDataFromServerRequest($request)->get('question_uuid')
+                    ),
+                    $DIC->assessment()->consumer()->revisionUuid(
+	                    new PostDataFromServerRequest($request)->get('revision_uuid')
+                    ),
+                    $user_id,
+                    json_encode(
+                        new PostDataFromServerRequest($request)->get('user_answer')
```

Comment:
So this implies that all question types (and plugin question types) provide the data in `$_POST['user_answer']`, and the question type itself is responsible for the data and structure of data in `$_POST['user_answer']`?

What if there are multiple questions on one page (not relevant for the current state of the test object in ILIAS, but relevant in future).

Reply:
> ... and the question type itself is responsible for the data and structure of data in $_POST['user_answer']

Yes!

> What if there are multiple questions on one page (not relevant for the current state of the test object in ILIAS, but relevant in future).

Thx for this question! Good Point!! 

I see two variants

Variant 1:
The consumer nevertheless sends the answers individually.  He could place a try/catch block outside a foreach. For this, however, the POST variable by rendering the question would have to be an array already, i.e. user_answer[].

In this case everything would be stored except those that fail.

Variant 2:
You only want to save the page if the whole page is saved. In this case we would have to offer another communication interface Post**Array**DataFromServerRequest.

Are there other variants? Should we already now provide the user_answer in the representation as an array?

Reply:
If ILIAS supports multiple questions on a page in future, a `Transaction` like handling (begin, commit, rollback) of the/all user answers could be maybe a requirement. So a 'everything succeeds or fails' behaviour could be the variant we should support.

But IMO both variants could be a valid scenario for the consumer of the question service. 

For the HTTP communication between (user) client and the consumer endpoint (e.g. UI controller of the Test repository object) the UUID of the question should be part of the structure of the `$_POST['user_answer']` array, e.g.:

```php
$_POST['user_answer']['d01a14f5-544d-418f-87cf-7c293c673531'] = [
    'gap1' => 'a² + 2ab + b²', // Some question specific user answer data
];
```

Reply:
In addition to Martins variants for solving the problems with the user_answer post variable,
Variant 3:

The consumer will define the post variable to be used by the question types. This way the consumer gets the full control over his own form fields' post variables.

Reply:
In general it will be technically possible to implement either a user answer save service that takes all user answers at once or a service that takes care about one user answer only.

The question is from high value, we will consider this in mind. I just need to say, the question also aims general conceptional questions, that needs to be clarified within the related feature request.

Even the offline implementation of questions supporting the test and also the learning module could have impact on this question.

Very Thanks @mjansenDatabay  (!) <3